lite/kernels/imagination_nna/bridges/conv_op.cc

```diff
@@ -0,0 +1,278 @@
+  CHECK(op_info->HasInputScale(filter_name));
+  std::vector<float> weight_scale = op_info->GetInputScale(filter_name);
+  CHECK(op_info->HasOutputScale(output_name));
+  float output_scale = op_info->GetOutputScale(output_name)[0];
+
+  TensorInfo qnt;
+
+  // Input node
+  std::shared_ptr<Node> input_node = nullptr;
+  imgdnn_tensor in_tensor;
+  if (graph->Has(input_name)) {
+    input_node = graph->Get(input_name);
+    in_tensor = input_node->data();
+  } else {
+    TensorInfoReset(&qnt);
+    qnt.type = IMGDNN_TYPE_Q_U8;
+    qnt.scales.push_back(input_scale);
+    qnt.zero_points.push_back(128);
+    input_node = graph->Add(input_name, *input, qnt, Node::Role::kInput);
+    in_tensor = input_node->data();
+  }
+
+  if (paddings.size() == 2L) {
+    for (size_t i = 0; i < strides.size(); ++i) {
+      int copy_pad = *(paddings.begin() + 2 * i);
+      paddings.insert(paddings.begin() + 2 * i + 1, copy_pad);
+    }
+  }
+  CHECK_EQ(paddings.size(), 4L)
+      << "[NNA] Paddings size should be the same or twice as the input size.";
+
+  std::string padding_algorithm("");
+  if (op_info->HasAttr("padding_algorithm")) {
+    padding_algorithm = op_info->GetAttr<std::string>("padding_algorithm");
+  }
+  operators::UpdatePaddingAndDilation(&paddings,
+                                      &dilations,
+                                      strides,
+                                      padding_algorithm,
+                                      input_dims,
+                                      filter_dims);
+
+  // Check depthwise mode, and decide whether use ConvolutionDepthwise Op
+  bool is_depthwise_mode = (ic == groups && oc == groups && groups != 1);
+
+  // Filter node
+  bool per_channel = isScalesPerChannel(weight_scale);
+  TensorInfoReset(&qnt);
+  uint8_t *weights_u8 =
+      graph->GetBuilder()->GetBufromPool(filter_dims.production());
+  char *weight_src = static_cast<char *>(filter->raw_data());
+
+  qnt.type = IMGDNN_TYPE_Q_U8;
+  if (per_channel) {
+    qnt.scales.assign(weight_scale.begin(), weight_scale.end());
+    qnt.zero_points.assign(weight_scale.size(), 128);
+    qnt.count = oc;
+    qnt.axis = 1;
+  } else {
+    qnt.scales.push_back(weight_scale.at(0));
+    qnt.zero_points.push_back(128);
+  }
+  for (int i = 0; i < filter_dims.production(); i++) {
+    weights_u8[i] = static_cast<uint8_t>(weight_src[i] + 128);
+  }
+
+  std::shared_ptr<Node> filter_node = graph->Add(filter_name,
+                                                 weights_u8,
+                                                 filter_dims.Vectorize(),
+                                                 qnt,
+                                                 Node::Role::kConst);
+  imgdnn_tensor filter_tensor = filter_node->data();
+
+  // Add bias node if exists bias
+  // Supports the bias nodes with the following dimensions
+  // 0: {oc}
+  // 1: {1, oc, oh, ow}
+  // 2: {n, oc, oh, ow}
+  std::shared_ptr<Node> bias_node = NULL;
```

Comment:
use c++ style `nullptr` instead of `NULL`

No need to initialize a shared_ptr.

Reply:
Done